### DISCUSSION TITLE
Lesson 6: TypeError: argument of type 'NoneType' is not iterable

### DISCUSSION THREAD

**Discussion by KayveliDevelopment** [smartcontractkit/full-blockchain-solidity-course-py#1172] [Q&A] [answered]
Hello guys I hope someone has a solution for this error as I've been stuck on it for the past day,  at around 5:31:40 on the solidity CodeCamp video I get a 'NoneType' error when deploying my contract. I have checked through the code and where my files are and I made sure there's no errors in it, I have also uninstalled and re-installed the ganache CLI, and I have also deleted the build folder and recompiled after adding the MockV3Aggregator and its still not working, if anyone has the solution It'll be much appreciated ^
![Screenshot_4](https://user-images.githubusercontent.com/100043020/155892532-e91e2fbf-d19b-4e1a-95af-4ed1784bab2b.png)
^

**Comment by n4n0b1t3** [comment]
it says something about loading the config. Can you post your config.yaml and .env file after masking private keys and such. I have seen this error a few times but cannot remember in which circumstances.

**Reply by KayveliDevelopment** [reply]
Hello, thanks for the reply! Here's my config and my .env file: 
![image](https://user-images.githubusercontent.com/100043020/155987459-da2fcf2b-3d8a-40cf-93bf-43af5fd2aa41.png)

**Reply by n4n0b1t3** [reply]
Hi @KayAAIC, can you try to explicitly state the network you want to use when calling brownie run. 
For example:
`(py39brownie) fund_me>brownie run .\scripts\deploy.py --network rinkeby`
I am pretty sure your default network is mainnet-fork and this is not set up. 

Something else is weird, your terminal did not change to ".\scripts\deploy.py" but keeps "scripts/deploy.py". But I don't know what this could be a symptom for...


**AND PLEASE ...** 

## Help others reproduce the problem
Not all questions benefit from including code, but if your problem is _with code_ you've written, you should include some. But don't just copy in your entire program! Not only is this likely to get you in trouble if you're posting for example your private key, it likely includes a lot of irrelevant details that readers will need to ignore when trying to reproduce the problem. Here are some guidelines:
- **DO NOT post images of code, data, error messages, etc.** - copy or type the text into the question. 
- Include just enough code to allow others to reproduce the problem.
- Always format your code by using the <> symbol of the editor.  First, copy and paste the code, then select the just pasted code and now click on the code format button <> of the editor. To make it even prettier you could add "python" after the first three backticks.

**Reply by KayveliDevelopment** [reply]
Hey @n4n0b1t3 thanks for the reply yet again, I was following the smart contract course step-by-step and in the tutorial he did not specify the network when I got this error and the ganache CLI was working fine until I added in the "MockV3Aggregator.sol" file, so I'm wondering if maybe there's an issue there.. I've setup my mainnet-fork to a local ganache blockchain following the tutorial also so I'm not sure. Do you recommend just uninstalling node, python, brownie and ganache-cli and reinstalling?

**Comment by cromewar** [comment]
Hello @KayAAIC  
Seems to be no problems in your code, please could you run `brownie compile` and paste here the output?

**Reply by KayveliDevelopment** [reply]
Sure thing! @cromewar 

```
PS C:\Users\Kaang\Demos\brownie_fund_me> brownie compile
INFO: Could not find files for the given pattern(s).
Brownie v1.18.1 - Python development framework for Ethereum

Compiling contracts...
  Solc version: 0.6.12
  Optimizer: Enabled  Runs: 200
  EVM Version: Istanbul
Generating build data...
 - smartcontractkit/chainlink-brownie-contracts@1.1.1/AggregatorInterface
 - smartcontractkit/chainlink-brownie-contracts@1.1.1/AggregatorV2V3Interface
 - smartcontractkit/chainlink-brownie-contracts@1.1.1/AggregatorV3Interface
 - smartcontractkit/chainlink-brownie-contracts@1.1.1/SafeMathChainlink
 - FundMe
 - MockV3Aggregator

Project has been compiled. Build artifacts saved at C:\Users\Kaang\demos\brownie_fund_me\build\contracts
```

**Reply by cromewar** [reply]
As you can see the `MockV3Aggregator` is compiling without problems so I don't think is a problem with that I'll review your code, so please paste here the entire deploy code as using the image above is hard to read.

**Reply by KayveliDevelopment** [reply]
sure sounds good, heres the code from deploy.py 
```
from brownie import FundMe, MockV3Aggregator, network, config
from scripts.helpful_scripts import (
    get_account,
    deploy_mocks,
    LOCAL_BLOCKCHAIN_ENVIROMENTS,
)


def deploy_fund_me():
    account = get_account()
    if network.show_active() not in LOCAL_BLOCKCHAIN_ENVIROMENTS:
        price_feed_address = config["networks"][network.show_active()][
            "eth_usd_price_feed"
        ]
    else:
        deploy_mocks()
        price_feed_address = MockV3Aggregator[-1].address

    fund_me = FundMe.deploy(
        price_feed_address,
        {"from": account},
        publish_source=config["networks"][network.show_active()].get("verify"),
    )
    print(f"Contract deployed to {fund_me.address}")


def main():
    deploy_fund_me()

```
and helpful_scripts.py..
```
from brownie import network, config, accounts, MockV3Aggregator
from web3 import Web3

LOCAL_BLOCKCHAIN_ENVIROMENTS = ["development", "ganache_local"]

DECIMALS = 18
STARTING_PRICE = 2000


def get_account():
    if network.show_active() in LOCAL_BLOCKCHAIN_ENVIROMENTS:
        return accounts[0]
    else:
        return accounts.add(config["wallets"]["from_key"])


def deploy_mocks():
    print(f"The network is {network.show_active()}")
    print("Deploying Mocks...")
    if len(MockV3Aggregator) <= 0:
        MockV3Aggregator.deploy(
            18, Web3.toWei(STARTING_PRICE, "ether"), {"from": get_account()}
        )
    print("Mocks Deployed!")

```

**Comment by n4n0b1t3** [comment]
please have a look at deploy_fund_me() lines 11-14, unless the default network is rinkeby the code is searching for a nonexisting key.

**Reply by KayveliDevelopment** [reply]
Okay I'll have a look but I did check several times that my code is correct, maybe due to a later version of brownie & ganache the code isn't compatible, perhaps I need to add something to do with the current gas price?

**Comment by n4n0b1t3** [comment]
OK, probably you are right, but what you can do to find out is:
```
(py39brownie) fund_me>brownie console
Brownie v1.18.1 - Python development framework for Ethereum

FundMeProject is the active project.

Launching 'ganache-cli.cmd --chain.vmErrorsOnRPCResponse true --wallet.totalAccounts 10 --hardfork istanbul --miner.blockGasLimit 12000000 --wallet.mnemonic brownie --server.port 8545'...
Brownie environment is ready.
>>> network.show_active()
'development'
>>>
```

**Reply by KayveliDevelopment** [reply]
I used the brownie console command and it returned the same error... :/
```
PS C:\Users\Kaang\Demos\brownie_fund_me> brownie console
INFO: Could not find files for the given pattern(s).
Brownie v1.18.1 - Python development framework for Ethereum

  File "C:\Users\Kaang\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\_cli\__main__.py", line 64, in main
    importlib.import_module(f"brownie._cli.{cmd}").main()
  File "C:\Users\Kaang\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\_cli\console.py", line 50, in main
    active_project.load_config()
  File "C:\Users\Kaang\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\project\main.py", line 462, in load_config
    _load_project_config(self._path)
  File "C:\Users\Kaang\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\_config.py", line 222, in _load_project_config
    and "cmd_settings" in values
TypeError: argument of type 'NoneType' is not iterable
```

**Reply by n4n0b1t3** [reply]
What is the path of your virtual environment, the one you call "eth-brownie"?
And what is your project directory path?

**Reply by KayveliDevelopment** [reply]
I changed my "eth-brownie" path following @cromewar fix and my project directory path is ` C:\Users\Kaang\demos\brownie_fund_me`  and the path for my virtual enviroment is  `c:\users\kaang\appdata\local\programs\python\python39\lib\site-packages`

**Reply by n4n0b1t3** [reply]
I am out of my depth here. Sorry I can't help. This path is weird through c:\users\kaang\appdata\local\programs\python\python39\lib\site-packages
Why would you install your virtual env here?

**Reply by KayveliDevelopment** [reply]
Its okay no worries, I've also tried looking on other articles and seems there isn't a known fix, It started happening right after I implemented MockV3Aggregator.sol, as for the virtual env path I didn't choose a custom install. I just installed through terminal on MSVC using `pip install eth-brownie virtualenv`

**Comment by cromewar** [comment]
Now I'm sure is something wrong with your brownie installation so please do the following:
1. Delete the brownie folder on `C:\Users\Kaang\.brownie` if you could not find it is ok
2. uninstall brownie with pipx
3. reinstall brownie using regular pip on a virtualenv (using python v3.8 or 3.9)

**Reply by KayveliDevelopment** [reply]
I followed your instructions and reinstalled brownie using regular pip on virtualenv on python 3.9
```
Installing collected packages: filelock, distlib, virtualenv, eth-brownie
Successfully installed distlib-0.3.4 eth-brownie-1.18.1 filelock-3.6.0 virtualenv-20.13.2
```
then i recompiled and attempted to run scripts and got the same error:
```
PS C:\Users\Kaang\Demos\brownie_fund_me> brownie run scripts/deploy.py 
INFO: Could not find files for the given pattern(s).
Brownie v1.18.1 - Python development framework for Ethereum

  File "c:\users\kaang\appdata\local\programs\python\python39\lib\site-packages\brownie\_cli\__main__.py", line 64, in main
    importlib.import_module(f"brownie._cli.{cmd}").main()
  File "c:\users\kaang\appdata\local\programs\python\python39\lib\site-packages\brownie\_cli\run.py", line 42, in main
    active_project.load_config()
  File "c:\users\kaang\appdata\local\programs\python\python39\lib\site-packages\brownie\project\main.py", line 462, in load_config
    _load_project_config(self._path)
  File "c:\users\kaang\appdata\local\programs\python\python39\lib\site-packages\brownie\_config.py", line 222, in _load_project_config
    and "cmd_settings" in values
TypeError: argument of type 'NoneType' is not iterable

```

**Reply by cromewar** [reply]
`_load_project_config(self._path)`

Could you please add the env variables manually on windows using the advanced configuration?

I've made more research on this and I think is not getting the env variables on your .env file.

**Reply by KayveliDevelopment** [reply]
I ended up finding a solution. I did this by git cloning the brownie_fundme project from https://github.com/PatrickAlphaC/brownie_fund_me in a new Folder, and then copying over my code, which ended up working. I'm not sure why but my guess is somehow the build files in the initial project were conflicting as I was messing around with uninstalling and reinstalling packages. ^^

**Comment by KayveliDevelopment** [comment, marked as answer]
UPDATE: I found a solution. I did this by git cloning the brownie_fundme project from https://github.com/PatrickAlphaC/brownie_fund_me in a new Folder, and then copying over my code, which ended up working. I'm not sure why but my guess is somehow the build files in the initial project were conflicting as I was messing around with uninstalling and reinstalling packages.

**Reply by cromewar** [reply]
strange error, but I'm glad is solved now.
